### PR TITLE
doc: add watchdog high-level design

### PR DESCRIPTION
**Pull request by dbkinder**
Review, transcode, and publish watchdog high-level design doc.

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

### CODE REVIEW

**Comment by hwang37** [comment]
I'm fine. No comment

File: doc/developer-guides/watchdog-hld.rst [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+:numref:`watchdog-device` shows the watchdog device workflow:
+
+.. figure:: images/watchdog-image2.png
+   :align: center
+   :width: 900px
+   :name: watchdog-device
+
+   Watchdog device flow
+
+The DM in the service OS (SOS) treats the watchdog as a passive device.
```
Comment (gvancuts) [inline doc/developer-guides/watchdog-hld.rst]: I think we spell Service OS with a capital S for service, I think we should stay consistent and use it here too.

File: doc/developer-guides/watchdog-hld.rst [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+It receives read/write commands from the watchdog driver, does the
+actions, and returns.  In ACRN, the commands are from user OS (UOS)
```
Comment (gvancuts) [inline doc/developer-guides/watchdog-hld.rst]: Same comment about using capital U for User OS

File: doc/developer-guides/watchdog-hld.rst [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+**********************
+
+When the UOS does a read or write operation on the watchdog device's
+registers or memory space (Port IO or Memory map I/O), it will trap into
+the hypervisor.   The hypervisor delivers the operation to the SOS/DM
+through IPI (inter-process interrupt) or shared memory, and the DM
+dispatches the operation to the watchdog emulation code.
+
+After the DM watchdog finishes emulating the read or write operation, it
+then calls ioctl to the SOS/kernel (acrn_vhm). VHM will call a
```
Comment (gvancuts) [inline doc/developer-guides/watchdog-hld.rst]: s/ioctl/``ioctl()``
s/acrn_vhm/``/dev/acrn_vhm``

File: doc/developer-guides/watchdog-hld.rst [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+hypercall to trap into the hypervisor to tell it the operation done, and
```
Comment (gvancuts) [inline doc/developer-guides/watchdog-hld.rst]: are we missing an "is" between "operation" and "done"?

File: doc/developer-guides/watchdog-hld.rst [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+   :width: 900px
+   :name: watchdog-workflow
+
+   Watchdog operation workflow
+
+Implementation in ACRN and how to use it
+****************************************
+
+In ACRN, the Intel 6300ESB watchdog device emulation is added into the
+DM PCI device tree. Its interface structure is (see ``pci_core.h``):
```
Comment (gvancuts) [inline doc/developer-guides/watchdog-hld.rst]: How about using the full path to that file in the source code, i.e. `devicemodel/include/pci_core.h`

Comment (gvancuts) [reply doc/developer-guides/watchdog-hld.rst]: It's actually not in that file at all... but I found it in another file [here](https://github.com/projectacrn/acrn-hypervisor/blob/master/devicemodel/hw/pci/wdt_i6300esb.c#L409-L417)

File: doc/developer-guides/watchdog-hld.rst [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+The main part in the watchdog emulation is the timer thread. It emulates
+the watchdog device timeout management. When it gets the kick action
+from the UOS, it resets the timer. If the timer expires before getting a
+timely kick action, it will call DM API to reboot that UOS.
+
+In the UOS launch script, add: ``-s xx,wdt-i6300esb`` into DM parameters.
+(xx is the virtual PCI BDF number as with other PCI devices)
+
+In UOS kernel, add a configure: ``CONFIG_I6300ESB_WDT=y``. After the UOS
```
Comment (gvancuts) [inline doc/developer-guides/watchdog-hld.rst]: How about rephrasing this a bit, to something like: "Make sure the UOS kernel has the I6300ESB driver enabled (`CONFIG_I6300ESB_WDT=y`)."

**Comment by mcao6** [comment]
OK, will improve this patch soon.

File: doc/developer-guides/watchdog-hld.rst [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+to monitor the system's running status, and resets the processor if the
+software crashes. In general,  hardware watchdogs rely on a piece of
+software running on the machine which must "kick" the watchdog device
+regularly, say every 10 seconds. If the watchdog doesn't get "kicked"
+after 60 seconds, for example, then the watchdog device asserts the
+RESET line which results in a hard reboot.
+
+For ACRN we emulate the watchdog hardware in the Intel 6300ESB chipset
+as a PCI device called 6300ESB watchdog and is added into the Device
+Model following follow the PCI device framework. The following
```
[the remark this answers is not in the document]
Comment (mcao6) [reply doc/developer-guides/watchdog-hld.rst]: @dbkinder could you help to modify it: "following the PCI ..."

File: doc/developer-guides/watchdog-hld.rst (hunk 1 above)
Comment (mcao6) [reply doc/developer-guides/watchdog-hld.rst]: @dbkinder could you change "service OS" to "Service OS"

File: doc/developer-guides/watchdog-hld.rst (hunk 2 above)
Comment (mcao6) [reply doc/developer-guides/watchdog-hld.rst]: @dbkinder , could you change "user OS" --> "User OS"

File: doc/developer-guides/watchdog-hld.rst (hunk 3 above)
Comment (mcao6) [reply doc/developer-guides/watchdog-hld.rst]: @gvancuts, sorry, not got your points here.

File: doc/developer-guides/watchdog-hld.rst (hunk 4 above)
Comment (mcao6) [reply doc/developer-guides/watchdog-hld.rst]: @dbkinder , could you help to add it "operation done" --> "operation is done"

File: doc/developer-guides/watchdog-hld.rst (hunk 5 above)
Comment (mcao6) [reply doc/developer-guides/watchdog-hld.rst]: the structure definition is in "pci_core.h", the watchdog used in "wdt_i6300esb.c", to me it is OK here, if the user need the code, they can find it for there are no duplicated definition.

File: doc/developer-guides/watchdog-hld.rst (hunk 6 above)
Comment (mcao6) [reply doc/developer-guides/watchdog-hld.rst]: @dbkinder , could you help to modify:  "In UOS kernel, add a configure: ``CONFIG_I6300ESB_WDT=y``" --> "Make sure the UOS kernel has the I6300ESB driver enabled (CONFIG_I6300ESB_WDT=y)."

**Pull request merged**